src/field_5x52_impl.h

```diff
@@ -498,4 +498,380 @@ static SECP256K1_INLINE void secp256k1_fe_from_storage(secp256k1_fe *r, const se
+    t[4] = e; t[5] = f; t[6] = g; t[7] = h;
+}
+
+static void secp256k1_fe_combine_2s(int64_t *t) {
+
+    int64_t a0 = t[ 0], a1 = t[ 1];
+    int64_t b0 = t[ 2], b1 = t[ 3];
+    int64_t c0 = t[ 4], c1 = t[ 5];
+    int64_t d0 = t[ 6], d1 = t[ 7];
+    int64_t e0 = t[ 8], e1 = t[ 9];
+    int64_t f0 = t[10], f1 = t[11];
+    int64_t g0 = t[12], g1 = t[13];
+    int64_t h0 = t[14], h1 = t[15];
+
+    secp256k1_fe_mul_add(e0, e1, a0, a1, f0, f1, c0, c1, &t[0]);
+    secp256k1_fe_mul_add(e0, e1, b0, b1, f0, f1, d0, d1, &t[4]);
+    secp256k1_fe_mul_add(g0, g1, a0, a1, h0, h1, c0, c1, &t[8]);
+    secp256k1_fe_mul_add(g0, g1, b0, b1, h0, h1, d0, d1, &t[12]);
+}
+
+static void secp256k1_fe_decode_matrix(secp256k1_fe *r, int64_t *t) {
+
+    uint64_t u0, u1, u2, u3, u4;
+    uint64_t r0, r1, r2, r3, r4;
+    int128_t cc;
+
+    cc  = t[0];
+    u0  = (uint64_t)cc; cc >>= 64;
+    cc += t[1];
+    u1  = (uint64_t)cc; cc >>= 64;
+    cc += t[2];
+    u2  = (uint64_t)cc; cc >>= 64;
+    cc += t[3];
+    u3  = (uint64_t)cc; cc >>= 64;
+    u4  = (uint64_t)cc;
+
+    VERIFY_CHECK(u4 == 0 || u4 == UINT64_MAX);
+
+    /* Add twice the field prime in case u4 is non-zero (which represents -2^256). */
+    r0 = 0xFFFFEFFFFFC2FULL * 2;
+    r1 = 0xFFFFFFFFFFFFFULL * 2;
+    r2 = 0xFFFFFFFFFFFFFULL * 2;
+    r3 = 0xFFFFFFFFFFFFFULL * 2;
+    r4 = 0x0FFFFFFFFFFFFULL * 2;
+
+    r0 += u0 & 0xFFFFFFFFFFFFFULL;
+    r1 += u0 >> 52 | ((u1 << 12) & 0xFFFFFFFFFFFFFULL);
+    r2 += u1 >> 40 | ((u2 << 24) & 0xFFFFFFFFFFFFFULL);
+    r3 += u2 >> 28 | ((u3 << 36) & 0xFFFFFFFFFFFFFULL);
+    r4 += u3 >> 16 |  (u4 << 48);
+
+    r->n[0] = r0;
+    r->n[1] = r1;
+    r->n[2] = r2;
+    r->n[3] = r3;
+    r->n[4] = r4;
+
+#ifdef VERIFY
+    r->magnitude = 2;
+    r->normalized = 0;
+    secp256k1_fe_verify(r);
+#endif
+}
+
+static void secp256k1_fe_encode_62(int64_t *r, const secp256k1_fe *a) {
+
+    const uint64_t M62 = UINT64_MAX >> 2;
+    const uint64_t *n = &a->n[0];
+    uint64_t a0 = n[0], a1 = n[1], a2 = n[2], a3 = n[3], a4 = n[4];
+
+#ifdef VERIFY
+    VERIFY_CHECK(a->normalized);
+#endif
+
+    r[0] = (a0       | a1 << 52) & M62;
+    r[1] = (a1 >> 10 | a2 << 42) & M62;
+    r[2] = (a2 >> 20 | a3 << 32) & M62;
+    r[3] = (a3 >> 30 | a4 << 22) & M62;
+    r[4] =  a4 >> 40;
+}
+
+static int secp256k1_fe_divsteps_62(uint16_t eta, uint64_t f0, uint64_t g0, int64_t *t) {
+
+    uint64_t u = -(uint64_t)1, v = 0, q = 0, r = -(uint64_t)1;
+    uint64_t c1, c2, f = f0, g = g0, x, y, z;
+    int i;
+
+    for (i = 0; i < 62; ++i) {
+
+        VERIFY_CHECK((u * f0 + v * g0) == -f << i);
+        VERIFY_CHECK((q * f0 + r * g0) == -g << i);
+
+        c1 = -(g & (eta >> 15));
+
+        x = (f ^ g) & c1;
+        f ^= x; g ^= x; g ^= c1; g -= c1;
+
+        y = (u ^ q) & c1;
+        u ^= y; q ^= y; q ^= c1; q -= c1;
+
+        z = (v ^ r) & c1;
+        v ^= z; r ^= z; r ^= c1; r -= c1;
+
+        eta = (eta ^ (uint16_t)c1) - (uint16_t)c1 - 1;
```

Comment:
On a related note, seeing integer types smaller than `int` makes me somewhat nervous. Most people looking at this line for example won't notice that all the arithmetic here is performed on (signed) `int` due to integer promotion rules. (I believe that's not an issue in this line, it's just a general remark.)

Reply:
Am I just missing a U i.e. 1U or else could you explain?

Reply:
The way integer operations (and this includes bitwise as well as unary operators,  shifts and comparisons) is done in C can be pretty unexpected for types smaller than `int`: Before any other rules apply, "Integer types smaller than `int` are promoted when an operation is performed on them. If all values of the original type can be represented as an `int`, the value of the smaller type is converted to an `int`; otherwise, it is converted to an unsigned `int`" [1]. On our targets, `int` is the same as `int32_t`, so an `uint16_t` for sure fits in an `int`. The possibility that an unsigned integer is converted to a signed integer is fucked up because signed overflow is undefined behavior.  And compilers have been reported to rely on this [2, 3]. 

For this line here, this means
`eta ^ (uint16_t)c1` is evaluated by first converting `eta` and `(uint16_t)c1` to `int` and then performing the `^`. Then for the `-`, the right operand `(uint16_t)c1`.  In that case,  it's easy to see that it doesn't make a difference, at least for our targets.

In general, if you ignore the theoretical cases like `int` 17 is bits wide, it seems (without a proof) that the only interesting operations are multiplication, left shifts, right shifts, and comparisons, the former two because of possible overflows, comparisons because of issues with comparing signed and unsigned values, and right shifts because it's relevant whether the left operand is signed or unsigned.

The real fuckup here is that the `stdint.h` types were supposed to be used for portable arithmetic but portable arithmetic is not possible in C because all integer arithmetic, no matter what type, implicitly depends on the size of `int` by the above rule. Our code would probably blow up on a target where `int` is 64 bits. (Yes, this exists.)

The typical trick to get around is to start the arithmetic with a unsigned int value, e.g., `(1U * eta) ^ (uint16_t)c1`. No need to point out that this is ugly.

[1] https://wiki.sei.cmu.edu/confluence/display/c/INT02-C.+Understand+integer+conversion+rules  
[2] https://stackoverflow.com/questions/24795651/whats-the-best-c-way-to-multiply-unsigned-integers-modularly-safely the link is for C++ but the same rules apply in this case
[3] https://www.cryptologie.net/article/418/integer-promotion-in-c/